### DISCUSSION TITLE
Enterprise security practices with AWS S3 Storage

### DISCUSSION THREAD

**Discussion by soundstep** [outline/outline#6256] [Ideas]
Hi,

We've created a self-hosted version of Outline (`v0.73.1`).

Note that the following is specific to our company. It is by no means a deal breaker for using Outline in most contexts.
I am simply reporting issues we had for your consideration.

The following is about AWS S3 security practices issues, related to two things:
- This single line of code:
https://github.com/outline/outline/blob/f494e28ce9b28be2cbb4c6d702658f53842fc830/server/models/helpers/AttachmentHelper.ts#L73
- The AWS public permissions: https://docs.getoutline.com/s/hosting/doc/file-storage-N4M0T6Ypu7#h-permissions

The `public-read` ACL is enforcing image uploads for avatars to be public. We are unsure why and would be happy to get more information. This ACL breached an "enterprise" security policy with our Cyber Security teams and other general Platform Engineering guidelines we have with AWS S3 encryption that is specific to us: we enforced AWS KMS encryption for S3 buckets. While KMS is the most secure encryption, a public ACL cannot work with KMS as you need a key to be able to read a file.
As a middle ground, we agreed internally to use AES256 to encrypt for this specific case.

We think this is also the reason why Outline requires the "Block all public access" S3 settings to be disabled. This is also a direct breach for us.

You may wonder why security measures would matter for an avatar? It does not matter too much directly, but in an enterprise world, Cyber Security will be applying blind general rules as much as they can.

To summarize, to get Outline working with S3, we had to:
- Disable the `Block all public access` settings on our file storage bucket as per Outline requirement.
- Switch to using AES256, rather than KMS-based encryption (if the AES256 encryption was not available, it would have been either a blocker for us, or make things much harder internally)

As far as our understanding goes, Outline allows us to secure the content of objects still, by generating pre-signed URLs for the browser to use when uploading and accessing content. However, it treats avatars differently from other kinds of uploads. It places a `public-read` ACL on them and tells the browser to make plain HTTP(S) requests for them, instead of using pre-signed URLs.

We need to, either:
- A: Make upstream changes to Outline, either:
	- To allow it to access avatars securely in the same way as other files
	- If there’s a good reason for Outline making avatars public, adding support for uploading these unencrypted (allowing us to default the bucket back to using KMS)
- B: Persuade CyberSecurity teams that either:
	- The current set-up is acceptable
	- It’s acceptable with mitigations:
		- Adding an IP-based bucket policy that restricts access to internal IPs
		- Something else?

Let me know what you think.

**Comment by tommoor** [comment]
The primary reason for this is that avatars are displayed on publicly shared pages (avatars include company logo and branding too). This allows allows them to be cached properly by the browser. Outline is primarily architected as a SaaS product and these types of profile pictures are generally public in this case. 

Do you require images to be behind authentication _in the product_ or is this just a matter of them being non-public on AWS even if the product allows an unauthenticated user to generated a signed url for download.

**Comment by soundstep** [comment]
> The primary reason for this is that avatars are displayed on publicly shared pages (avatars include company logo and branding too). This allows allows them to be cached properly by the browser. Outline is primarily architected as a SaaS product and these types of profile pictures are generally public in this case.

I see, it would be good to have an option of "disabling" the `public-read` using an environment variable to remove the public access if needed. Is that worth considering to cover all options?

> Do you require images to be behind authentication _in the product_ or is this just a matter of them being non-public on AWS even if the product allows an unauthenticated user to generated a signed url for download.

In our case, we don't need avatars to be public, we will not share URLs. It is only to match a company policy of having nothing public in the infrastructure instead of having exceptions left and right. This is something Cyber Security teams can impose if they don't have the capacity or will to manage exceptions.

It may feel like an isolated case, however, it may not be too hard to have the option of not having them public? (it may be an oversight)

**Reply by tommoor** [reply]
Yep, the option seems reasonable and not too much maintenance burden. It would need to disable all public sharing, and the "public branding" option as well and would be set in the server environment config.
